src/newrelic/collector/client.go

```diff
@@ -6,310 +6,331 @@
-	"time"
-
-	"golang.org/x/net/proxy"
-
-	"newrelic/log"
-	"newrelic/version"
+    "crypto/tls"
+    "encoding/json"
+    "fmt"
+    "io/ioutil"
+    "net"
+    "net/http"
+    "net/url"
+    "strings"
+    "time"
+
+    "golang.org/x/net/proxy"
+
+    "newrelic/log"
+    "newrelic/version"
 )
 
-var ErrPayloadTooLarge = errors.New("payload too large")
-var ErrUnauthorized = errors.New("unauthorized")
-var ErrUnsupportedMedia = errors.New("unsupported media")
-
 type CollectibleFunc func(auditVersion bool) ([]byte, error)
 
 func (fn CollectibleFunc) CollectorJSON(auditVersion bool) ([]byte, error) {
-	return fn(auditVersion)
+    return fn(auditVersion)
 }
 
 type Collectible interface {
-	CollectorJSON(auditVersion bool) ([]byte, error)
+    CollectorJSON(auditVersion bool) ([]byte, error)
 }
 
-type Cmd struct {
-	Name              string
-	Collector         string
-	License           LicenseKey
-	RunID             string
-	AgentLanguage     string
-	AgentVersion      string
-	Collectible       Collectible
+// RpmCmd contains fields specific to an individual call made to RPM.
+type RpmCmd struct {
+    Name              string
+    Collector         string
+    RunID             string
+    Data              []byte
+    License       LicenseKey
 	RequestHeadersMap map[string]string
+}
+
+// RpmControls contains fields which will be the same for all calls made
+// by the same application.
+type RpmControls struct {
+    Collectible   Collectible
+    AgentLanguage string
+    AgentVersion  string
+    ua            string
+}
+
+// RPMResponse contains a NR endpoint response.
+//
+// Agent Behavior Summary:
+//
+// on connect/preconnect:
+//     410 means shutdown
+//     200, 202 mean success (start run)
+//     all other response codes and errors mean try after backoff
+//
+// on harvest:
+//     410 means shutdown
+//     401, 409 mean restart run
+//     408, 429, 500, 503 mean save data for next harvest
+//     all other response codes and errors discard the data and continue the current harvest
+type RPMResponse struct {
+    StatusCode int
+    Body       []byte
+    // Err indicates whether or not the call was successful: newRPMResponse
+    // should be used to avoid mismatch between StatusCode and Err.
+    Err                      error
+    disconnectSecurityPolicy bool
+}
 
-	ua                string
+func newRPMResponse(StatusCode int) RPMResponse {
+    var err error
+    if StatusCode != 200 && StatusCode != 202 {
+        err = fmt.Errorf("response code: %d: %s", StatusCode, GetStatusCodeMessage(StatusCode))
+    }
+    return RPMResponse{StatusCode: StatusCode, Err: err}
+}
+// IsDisconnect indicates that the agent should disconnect.
+func (resp RPMResponse) IsDisconnect() bool {
+    return resp.StatusCode == 410 || resp.disconnectSecurityPolicy
+}
+// IsRestartException indicates that the agent should restart.
+func (resp RPMResponse) IsRestartException() bool {
+    return resp.StatusCode == 401 || resp.StatusCode == 409
```

Comment:
Should 401 still map to a restart for the PHP agent?

Reply:
I decided to leave it there to reduce future confusion and conform with the spec. There is a comment explaining where we differ from spec behavior.

Reply:
Maybe add a comment explaining that here as well - currently it looks like we consider a 401 a restart (following the spec) in this function while elsewhere the actual behavior differs from the spec.